### DISCUSSION TITLE
Aligning BSS data to page boundaries

### DISCUSSION THREAD

**Discussion by nrivard** [cc65/cc65#1669] [Q&A] [answered]
I'm having a devil of a time aligning specific global data variables to page boundaries. Because these are large buffers ($80 and $200 bytes in size) they should clearly be aligned to page boundaries. I tried using `.align` but that actually does it for the entire segment. I tried overriding relocatable as well but that was a mess. Is the standard solution that i have to declare _per variable_ segments to accomplish this?

Consider:
```asm
.struct SDCardDataPacket
    data        .res 512                ; this is first in memory to align with page boundary
    token       .byte                   ; this will actually get sent first but comes after `data` in memory
    crc         .res 2                  ; needs to be fetched even if ignored
.endstruct

;; how do i guarantee this starts on a page boundary?
ReceivedDataPacket: .tag SDCardDataPacket
```

**Comment by mrdudz** [comment, marked as answer]
It very much depends on what exactly you are doing, and what the constraints of your platform are. Generally there are some things to consider:
- .align will align data relative to the segment start. this is very unintuitive and unlike every other assembler i have seen :)
- .align will "skip forward" until the aligned address is found, creating a "hole". the linker will not fill those holes, segments are simply created front to back in the link order

What i did for this kind of thing was creating a page aligned seperate segment, which then can be used for the arrays that benefit from being page aligned.

**Comment by nrivard** [comment]
@mrdudz Ok that makes sense. It's not ideal because then my data isn't contiguous but that's kind of the name of the game in 6502 world.

**Comment by mrdudz** [comment]
You can reduce the problem somewhat by doing the following:
- make sure the .bss segment is page aligned
- put all arrays that need alignment first into that segment
- followed by all other data

**Comment by groessler** [comment]
It's in general a good habit to sort e.g. C/C++ struct members by alignment. Strictest alignment first, then descending. E.g.

- 128bit vars
- 64bit vars
- 32bit vars
- 16bit vars
- 8bit vars

In that order. then minimum space is wasted for padding.

**Comment by greg-king5** [comment]
mrdudz's advice means that you should avoid structures like your example.  Instead, you can write your variables like this:
```asm
.bss
        .align  $0100
ReceivedDataBuffer:
        .res    $0200
SendDataBuffer:
        .res    $0200

; ... other aligned arrays ...

ReceivedToken:
        .res    1
ReceivedCRC:
        .res    2
SendToken:
        .res    1
SendCRC:
        .res    2
```

**Reply by nrivard** [reply]
Yep understood. Wrote it less clearly as
> It's not ideal because then my data isn't contiguous

which honestly does kind of suck, but then pagesize, register width, and other aspects make 6502 less suited for C like structures.

**Reply by nrivard** [reply]
Actually though, question: if `.bss` is page aligned, wouldn't _each_ of those non-page sized variables get their own page by the linker?
```asm
.bss ; starts at $0200 for this example
    .align $0100

ReceivedDataBuffer: .res $0200         ; $0200..$03FF
ReceivedToken: .res 2                  ; $0400
ReceivedCRC:   .res 1                  ; $0500 ??? or is it $0402?
```

**Comment by mrdudz** [comment]
> which honestly does kind of suck, but then pagesize, register width, and other aspects make 6502 less suited for C like structures.

Perhaps in this case an alternative solution would be putting a pointer into the struct rather than an array?

> Actually though, question: if .bss is page aligned, wouldn't each of those non-page sized variables get their own page by the linker?

No. Any .align directive is basically a single "fast forward" to the next address that matches the alignment. Its not an attribute that sticks to every object that comes after that. basically what it does is
```
while (offset % alignment != 0) { .byte fillvalue }
```
